File: Grafo/uri1799.py
import heapq

class Graph:
	def __init__(self, v):
		self.v = v
		self.matrix = []

		for i in range(v):
			lista = []
			for j in range(v):
				lista.append(0)
			self.matrix.append(lista)

	def addEdge(self, u, v):
		self.matrix[u][v] = self.matrix[v][u] = 1


	def dijkstra(self, de, para):
		distance = [4004] * self.v
		distance[de] = 0
		queue = []
		heapq.heappush(queue, (0, de))
		while(queue):
			weight, u = heapq.heappop(queue)
			#print(str(weight) + " " + str(u))
			for i in range(self.v):
				if(self.matrix[u][i] == 1 and distance[i] > weight + 1):
					distance[i] = weight + 1
					heapq.heappush(queue, (distance[i], i))
		return distance[para]


v, e = map(int, input().split())
g = Graph(v)
dictionary = {}
index = 0
for i in range(e):
	u, v = input().split()
	if(dictionary.get(u, -1) == -1):
		dictionary[u] = index
		index += 1
	if(dictionary.get(v, -1) == -1):
		dictionary[v] = index
		index += 1
	g.matrix[dictionary[u]][dictionary[v]] = g.matrix[dictionary[v]][dictionary[u]] = 1

print(g.dijkstra(dictionary['Entrada'], dictionary['*']) + g.dijkstra(dictionary['*'], dictionary['Saida']))

File: Grafo/uri1799.cpp
#include <bits/stdc++.h>

using namespace std;

typedef vector<vector<int>> Edge;

struct Graph{
	int v;
	Edge e;

	Graph(int v){
		this->v = v;
		e.assign(v, vector<int>());
	}

	int dijkstra(int start, int end){
		int *distance = new int[v];
		for(int i = 0; i < v; i++){
			distance[i] = 4004;
		}

		distance[start] = 0;
		priority_queue<pair<int, int>> pq;
		pq.push({0, start});

		while(!pq.empty()){
			pair<int, int> par = pq.top();
			pq.pop();
			int u, w;
			u = par.second;
			w = par.first;

			for(auto i = e[u].begin(); i != e[u].end(); i++){
				if(distance[*i] > w + 1){
					distance[*i] = w + 1;
					pq.push({distance[*i], *i});
				}
			}
		} 
		return distance[end];
	}
};

int main()
{
	int v, e;
	scanf("%d %d", &v, &e);
	Graph g(v);

	map<string, int> mapa;
	int index = 0;
	getchar();
	while(e--){
		string u, v;
		string temp;
		getline(cin, temp);
		u = temp.substr(0, temp.find(' '));
		v = temp.substr(temp.find(' ') + 1, temp.size());
		//cout << u << v << endl;
		if(mapa.find(u) == mapa.end()){	
			mapa[u] = index++;
		}
		if(mapa.find(v) == mapa.end()){
			mapa[v] = index++;
		}

		g.e[mapa[u]].push_back(mapa[v]);
		g.e[mapa[v]].push_back(mapa[u]);
	}
	printf("%d\n", g.dijkstra(mapa["Entrada"], mapa["*"]) + g.dijkstra(mapa["*"], mapa["Saida"]));
}